docs: Add a user guide on ScrapeConfig

xiu opened the pull request:
## Description

Related: #5607 

## Type of change

_What type of changes does your code introduce to the Prometheus operator? Put an `x` in the box that apply._

- [ ] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [ ] `FEATURE` (non-breaking change which adds functionality)
- [ ] `BUGFIX` (non-breaking change which fixes an issue)
- [ ] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [x] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Changelog entry

n/a

slashpai commented on Documentation/user-guides/scrapeconfig.md:
```diff
@@ -0,0 +1,132 @@
+    - labels:
+        job: prometheus
+      targets:
+      - aremoteprometheus.io:9090
```
we could use prometheus.demo.do.prometheus.io

slashpai commented on Documentation/user-guides/scrapeconfig.md:
```diff
@@ -0,0 +1,132 @@
+  name: your-prometheus
+  namespace: monitoring
+spec:
+  [...]
```
probably add configmap section here?

```yaml
spec:
  configMaps:
   - scrape-file-sd-targets
```

slashpai replied:
Also we would need to add `scrapeConfigSelector: {}`. It doesn't work without adding that

slashpai commented on Documentation/user-guides/scrapeconfig.md:
```diff
@@ -0,0 +1,132 @@
+spec:
+  staticConfigs:
+    - targets: 
+      - 10.1.2.3:9100
```
add label too?

slashpai reviewed (commented):
Just a few comments :)

I had created similar example in here https://github.com/slashpai/prometheus-operator-examples/tree/main/scrape_config for presentation at KCD

xiu commented:
@slashpai Thanks! Updated and added `kubernetes_sd`, please check the example when you get a chance. 

As for your examples, should we include them in the prometheus-operator repo?

slashpai commented on Documentation/user-guides/scrapeconfig.md:
```diff
@@ -0,0 +1,162 @@
+  labels:
+    app.kubernetes.io/name: scrape-config-example
+spec:
+  kubernetesSDConfig:
```
We would need a TLSConfig and Authorization section SC spec 
here is the example I tested https://github.com/prometheus-operator/prometheus-operator/pull/5675#issuecomment-1608867483

Also cluster role binding was required for prometheus so that prometheus serviceAccount has access to list Node objects otherwise prometheus errors about access issue so we should mention that as well

xiu replied:
Following I will remove `kubernetes_sd` to keep it short and we can figure out a specific documentation for it as there is more context than that (e.g. "How to scrape an exporter located on all the kubernetes worker").

ArthurSens commented on Documentation/user-guides/scrapeconfig.md:
```diff
@@ -0,0 +1,162 @@
+
+# Use ScrapeConfig to scrape an external target
+
+`ScrapeConfig` currently supports a limited set of service discoveries:
+* `static_config`
+* `file_sd`
+* `http_sd`
+* `kubernetes_sd`
```
Do we plan to document all fields added to ScrapeConfig?

I'm a bit afraid that we could easily miss documentation as time goes on. Providing examples for all service discovery options might make this page quite big 🤔

xiu replied:
I was actually thinking about that and wanted to discuss having some use case based documentation rather than this (e.g. how to scrape exporters on the kubernetes nodes or how to scrape an external target etc). That way we don't have to document everything in one page but can focus on use cases that drove the implementation.

What do you think?

ArthurSens replied:
Since we don't have any documentation and people are already asking for guidance on ScrapeConfig, we could merge while the CRD is still small. We could rethink and rewrite it later.


I was raising this question mostly because I was concerned that we would have to have this discussion at some point 🤔

simonpasquier replied:
I'd be +1 on having the documentation around specific use cases. And I feel that the current examples fulfill the requirement without going into too many details.

xiu replied:
Alright, I'll keep it at that for now and we can figure out use case based docs as we go 👍

simonpasquier commented on Documentation/user-guides/scrapeconfig.md:
```diff
@@ -0,0 +1,162 @@
+    - labels:
+        job: prometheus
+      targets:
+        - prometheus.demo.do.prometheus.io:9090
```
I like that we use the Prometheus demo instance as it's working out of the box. Do we need the other `10.1.2.3:9100` target?

xiu replied:
We don't, removed it 👍

xiu commented on Documentation/user-guides/scrapeconfig.md:
```diff
@@ -0,0 +1,162 @@
+  fileSDConfigs:
+    - files:
+        - /etc/prometheus/configmaps/scrape-file-sd-targets/targets.yaml
+      refreshInterval: 15s
```
done 👍

xiu commented:
As discussed in previous reviews, I kept examples short and we'll focus on more targetted documentations when needed (for instance, one on how to scrape node-exporter running outside kubernetes on all the kubernetes workers). As such, the doc focusses on the basic sd mechanisms: file, http, static.

simonpasquier commented on Documentation/user-guides/scrapeconfig.md:
```diff
@@ -0,0 +1,137 @@
+
+## `static_config`
+
+For example, to scrape a target located at `http://10.1.2.3:9100`, use the following:
```
```suggestion
For example, to scrape the target located at `http://prometheus.demo.do.prometheus.io:9090`, use the following:
```

simonpasquier commented on Documentation/user-guides/scrapeconfig.md:
```diff
@@ -0,0 +1,137 @@
+apiVersion: coreos.monitoring.com/v1alpha1
+kind: ScrapeConfig
+metadata:
+  name: my-scrape-config
```
(nit)
```suggestion
  name: static-config
```

simonpasquier commented on Documentation/user-guides/scrapeconfig.md:
```diff
@@ -0,0 +1,137 @@
+* `prometheus-operator` `>v0.65.1`
+* `ScrapeConfig` installed in the cluster. Make sure to restart the operator after having pushed the CRD.
+
+## Configure Prometheus or PrometheusAgent to select ScrapeConfigs
```
(nit) 
```suggestion
# Configure Prometheus or PrometheusAgent to select ScrapeConfigs
```

xiu replied:
This was for me part of the prerequisites but I'm not too fussed either way

simonpasquier commented on Documentation/user-guides/scrapeconfig.md:
```diff
@@ -0,0 +1,137 @@
+apiVersion: monitoring.coreos.com/v1alpha1
+kind: ScrapeConfig
+metadata:
+  name: scrape-config-example
```
```suggestion
  name: file-sd
```

xiu commented:
@simonpasquier adressed your comments, can you take another look?

simonpasquier reviewed (approved):
Thanks! :+1:

xiu commented:
@slashpai are you good with the last iterations? if so, I'll merge

slashpai commented:
yes lgtm :)
